### PR TITLE
API Backend refactor

### PR DESCRIPTION
**Pull request by MuawizChaudhary**
This PR is in progress. All the tests currently do not run. 

Refactors the named tuple convention out of the 1d.

### CODE REVIEW

File: tests/scattering1d/test_tensorflow_backend_1d.py [hunk 1]
```diff
@@ -1,7 +1,7 @@
 import pytest
 import numpy as np
-
-from kymatio.scattering1d.backend.tensorflow_backend import backend
```
Comment (janden) [inline tests/scattering1d/test_tensorflow_backend_1d.py]: No changes here.

File: tests/scattering1d/test_torch_backend_1d.py [hunk 2]
```diff
@@ -12,13 +12,14 @@
         skcuda_available = True
 except:
     Warning('torch_skcuda backend not available.')
+skcuda_available = False
```
Comment (janden) [inline tests/scattering1d/test_torch_backend_1d.py:15]: Why? Where's the `torch_skcuda` backend anyhow?

File: tests/scattering1d/test_torch_backend_1d.py [hunk 3]
```diff
@@ -176,6 +177,8 @@ def test_unpad():
     # test unpading of a random tensor
     x = torch.randn(8, 4, 1)
 
+    Torch1DBackend()
```
Comment (janden) [inline tests/scattering1d/test_torch_backend_1d.py:180]: What's going on here?

File: kymatio/backend/numpy_backend.py [hunk 4]
```diff
@@ -1,24 +1,35 @@
...
+    if not _is_complex(x):
+        raise TypeError('The input should be complex.')
+
+
+def real_check(x):
+    if not _is_real(x):
+        raise TypeError('The input should be real.')
+
+
+def _is_complex(x):
```
Comment (janden) [inline kymatio/backend/numpy_backend.py:19]: Why are these moved outside? Don't they depend on the `np` module we're using (regular NumPy vs. Jax)?

File: kymatio/backend/tensorflow_backend.py [hunk 5]
```diff
@@ -1,49 +1,54 @@
...
+    return (x.dtype == np.float32) or (x.dtype == np.float64)
+   
+ 
+class TensorFlowBackend():
+    def __init__(self):
+        self.name = 'tensorflow'
+        self.modulus = Modulus()
+  
+
+    def concat(self, arrays, dim):
```
Comment (janden) [inline kymatio/backend/tensorflow_backend.py:50]: Why not `concatenate` like we have elsewhere?

File: kymatio/backend/torch_backend.py [hunk 6]
```diff
@@ -3,163 +3,167 @@
...
+
+
+def input_checks(x):
+    if x is None:
+        raise TypeError('The input should be not empty.')
+
+    contiguous_check(x)
+
+
+def complex_check(x):
```
Comment (janden) [inline kymatio/backend/torch_backend.py:134]: Why are these not in the class?

File: kymatio/scattering1d/backend/jax_backend.py [hunk 7]
```diff
@@ -0,0 +1,5 @@
+
```
Comment (janden) [inline kymatio/scattering1d/backend/jax_backend.py:1]: Are we including the Jax backend in this PR?

File: kymatio/scattering1d/backend/numpy_backend.py [hunk 8]
```diff
@@ -1,12 +1,12 @@
 # Authors: Edouard Oyallon, Joakim Anden, Mathieu Andreux
 BACKEND_NAME = 'numpy'
 
-from ...backend.numpy_backend import NumpyBaseBackend
+from ...backend.numpy_backend import NumpyBackend, complex_check, real_check
```
Comment (janden) [inline kymatio/scattering1d/backend/numpy_backend.py:4]: If checks are included in the class, we wouldn't need these extra imports here.

File: tests/scattering1d/test_torch_backend_1d.py [hunk 9]
```diff
@@ -173,19 +173,22 @@ def test_subsample_fourier(backend, device, random_state=42):
...
 
 
-def test_unpad():
-    # test unpading of a random tensor
-    x = torch.randn(8, 4, 1)
+@pytest.mark.parametrize("backend", backends)
+@pytest.mark.parametrize("device", devices)
+def test_unpad(backend, device):
+    if backend.name == "torch_skcuda" and device == "cpu":
+        pytest.skip()
```
Comment (janden) [inline tests/scattering1d/test_torch_backend_1d.py:180]: Why are we adding this here? Can avoid these skips using the backend–device tuple scheme.

File: tests/scattering1d/test_torch_scattering1d.py [hunk 10]
```diff
@@ -13,7 +13,7 @@
     if torch.cuda.is_available():
         from skcuda import cublas
         import cupy
-        skcuda_available = True
+#        skcuda_available = True
```
Comment (janden) [inline tests/scattering1d/test_torch_scattering1d.py:16]: ?

File: kymatio/backend/numpy_backend.py (hunk 4 above)
Comment (MuawizChaudhary) [reply kymatio/backend/numpy_backend.py:19]: We use them in the frontend. Should I just use the back end variable to call the appropriate function in the frontend?

File: kymatio/backend/tensorflow_backend.py (hunk 5 above)
Comment (MuawizChaudhary) [reply kymatio/backend/tensorflow_backend.py:50]: so we call this function in the other files. 

before, when we had the namedtuple convention we would be able to do this:
`backend.concatenate = lambda x: concatenate(x, -2)`

Now, from the inheritance of this backend here, we have to worry about this:

`def concatenate(self, x):
    return self.concatenate(x, -2)`

this makes the code a bit scary, I don't know whats going on, additionally it was crashing with this convention I described above.

File: kymatio/backend/torch_backend.py (hunk 6 above)
Comment (MuawizChaudhary) [reply kymatio/backend/torch_backend.py:134]: called in frontend

File: kymatio/backend/torch_backend.py [hunk 11]
```diff
@@ -1,230 +1,242 @@
...
+    def contiguous_check(self, x):
+        if not x.is_contiguous():
+            raise RuntimeError('Tensors must be contiguous.')
+    
 
-    if B.device.type == 'cuda':
-        if A.device.type == 'cuda':
-            if A.device.index != B.device.index:
-                raise TypeError('Input and filter must be on the same GPU.')
+    def cdgmm(self, A, B, inplace=False):
```
[the remark this answers is not in the document]
Comment (MuawizChaudhary) [reply kymatio/backend/torch_backend.py:168]: that was another PR that hasn't been merged yet.

File: kymatio/scattering1d/backend/jax_backend.py (hunk 7 above)
Comment (MuawizChaudhary) [reply kymatio/scattering1d/backend/jax_backend.py:1]: My bad.

File: kymatio/scattering1d/backend/numpy_backend.py (hunk 8 above)
Comment (MuawizChaudhary) [reply kymatio/scattering1d/backend/numpy_backend.py:4]: True.

File: kymatio/scattering1d/backend/numpy_backend.py [hunk 12]
```diff
@@ -1,115 +1,126 @@
...
-    """
-    if (pad_left >= x.shape[-1]) or (pad_right >= x.shape[-1]):
-        raise ValueError('Indefinite padding size (larger than tensor).')
+from ...backend.numpy_backend import NumpyBaseBackend
+
+
+class Numpy1DBackend(NumpyBaseBackend):
+    def __init__(self, np):
+        super(Numpy1DBackend, self).__init__(np)
+        self.np = np
```
[the remark this answers is not in the document]
Comment (MuawizChaudhary) [reply kymatio/scattering1d/backend/numpy_backend.py:10]: Correct.

File: kymatio/scattering1d/backend/torch_backend.py [hunk 13]
```diff
@@ -3,129 +3,122 @@
...
+from ...backend.torch_backend import TorchBaseBackend
 
 
-def subsample_fourier(x, k):
-    """Subsampling in the Fourier domain
+class Torch1DBackend(TorchBaseBackend):
+    def __init__(self):
+        super(Torch1DBackend, self).__init__()
+        self.name = 'torch'
+        pass
```
[the remark this answers is not in the document]
Comment (MuawizChaudhary) [reply kymatio/scattering1d/backend/torch_backend.py:14]: not needed.

File: tests/scattering1d/test_torch_scattering1d.py (hunk 10 above)
Comment (MuawizChaudhary) [reply tests/scattering1d/test_torch_scattering1d.py:16]: Just temporary.

File: tests/scattering1d/test_torch_backend_1d.py (hunk 2 above)
Comment (MuawizChaudhary) [reply tests/scattering1d/test_torch_backend_1d.py:15]: I'll be pushing it soon enough.

File: kymatio/backend/torch_backend.py [hunk 14]
```diff
@@ -1,230 +1,242 @@
...
+            Returns
+            -------
+            output : tensor
+                A tensor with the same dimensions as x, such that output[..., 0]
+                contains the complex modulus of x, while output[..., 1] = 0.
+        """
+        def __call__(self, x):
+            complex_contiguous_check(x)
+            norm = modulus(x)[..., None]
+            return norm
```
[the remark this answers is not in the document]
Comment (MuawizChaudhary) [reply kymatio/backend/torch_backend.py]: I'll create an issue!

File: kymatio/backend/torch_backend.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (MuawizChaudhary) [reply kymatio/backend/torch_backend.py:168]: Actually you are right.

[the remark this answers is not in the document]
Comment (MuawizChaudhary) [reply kymatio/backend/torch_backend.py:168]: I based this PR based off of the 1d test pr. The 1d PR had not been rebased ontop of the change. How do we want to proceed?

File: kymatio/backend/numpy_backend.py (hunk 4 above)
Comment (eickenberg) [reply kymatio/backend/numpy_backend.py:19]: I don't understand this answer or its implications.

I guess the question is whether jax can handle numpy dtype objects or not

File: kymatio/backend/numpy_backend.py [hunk 15]
```diff
@@ -1,78 +1,87 @@
...
-        raise TypeError('The second input must be complex or real.')
-
-    if inplace:
-        return np.multiply(A, B, out=A)
-    else:
-        return A * B
+    return (x.dtype == numpy.float32) or (x.dtype == numpy.float64)
+
+    
+class NumpyBackend():
```
Comment (eickenberg) [inline kymatio/backend/numpy_backend.py:27]: parentheses not necessary

File: kymatio/backend/numpy_backend.py [hunk 16]
```diff
@@ -1,78 +1,87 @@
...
+    def __init__(self, np):
+        self.np = np
+        self.name = 'numpy'
```
Comment (eickenberg) [inline kymatio/backend/numpy_backend.py:30]: so: The numpy object is passed in, but the `self.name` will be inherited?

will we also be having a cupy backend then?

File: kymatio/backend/tensorflow_backend.py [hunk 17]
```diff
@@ -1,23 +1,5 @@
...
-    if not _is_real(x):
-        raise TypeError('The input should be real.')
-
-def _is_complex(x):
-    return (x.dtype == np.complex64) or (x.dtype == np.complex128)
-
-
-def _is_real(x):
-    return (x.dtype == np.float32) or (x.dtype == np.float64)
+import tensorflow as tf
```
Comment (eickenberg) [inline kymatio/backend/tensorflow_backend.py:2]: Do we need to put the tf backend into a class? What inherits from it? Can't it just stay as is?

File: kymatio/backend/tensorflow_backend.py (hunk 5 above)
Comment (eickenberg) [reply kymatio/backend/tensorflow_backend.py:50]: I don't follow

File: kymatio/backend/tensorflow_backend.py [hunk 18]
```diff
@@ -41,36 +23,60 @@ def __call__(self, x):
...
 
-    if not _is_complex(B) and not _is_real(B):
-        raise TypeError('The second input must be complex or real.')
 
-    return A * B
+def _is_real(x):
+    return (x.dtype == np.float32) or (x.dtype == np.float64)
+   
+ 
+class TensorFlowBackend():
```
Comment (eickenberg) [inline kymatio/backend/tensorflow_backend.py:44]: parentheses not necessary

File: kymatio/backend/torch_backend.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (eickenberg) [reply kymatio/backend/torch_backend.py:168]: such that the end result is that there is no more `inplace` stuff. I would suggest removing it here. Otherwise it might end up showing up again when this gets merged

File: kymatio/scattering1d/backend/jax_backend.py [hunk 19]
```diff
@@ -0,0 +1,5 @@
...
+from .numpy_backend import NumpyBackend1D
+import jax.numpy 
+
+backend = NumpyBackend1D(jax.numpy)
```
Comment (eickenberg) [inline kymatio/scattering1d/backend/jax_backend.py:5]: This thing, according to what I saw above will still have `self.name` be `numpy`

File: kymatio/scattering1d/backend/numpy_backend.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (eickenberg) [reply kymatio/scattering1d/backend/numpy_backend.py:10]: can't we pass in the numpy/jax object via inheritance?

File: kymatio/scattering1d/backend/numpy_backend.py [hunk 20]
```diff
@@ -1,115 +1,126 @@
...
+        -------
+        res : tensor
+            The input tensor periodized along the next to last axis to yield a
+            tensor of size x.shape[-2] // k along that dimension.
+        """
+        self.complex_check(x)
+    
+        y = x.reshape(-1, k, x.shape[-1] // k)
+    
+        res = y.mean(axis=(-2,))
```
[the remark this answers is not in the document]
Comment (eickenberg) [reply kymatio/scattering1d/backend/numpy_backend.py:37]: `axis=-2`

File: kymatio/scattering1d/backend/tensorflow_backend.py [hunk 21]
```diff
@@ -1,109 +1,104 @@
...
-backend.pad = pad
-backend.rfft = rfft
-backend.irfft = irfft
-backend.ifft = ifft
-backend.concatenate = lambda x: concatenate(x, -2)
+import tensorflow as tf
+from ...backend.tensorflow_backend import TensorFlowBackend, complex_check, real_check
+
+
+class TensorFlowBackend1D(TensorFlowBackend):
```
Comment (eickenberg) [inline kymatio/scattering1d/backend/tensorflow_backend.py:7]: again, why do we need this for tf?

File: kymatio/scattering1d/backend/tensorflow_backend.py [hunk 22]
```diff
@@ -1,109 +1,104 @@
...
+        -------
+        res : tensor
+            The input tensor periodized along the next to last axis to yield a
+            tensor of size x.shape[-2] // k along that dimension.
+        """
+        complex_check(x)
+    
+        y = tf.reshape(x, (-1, k, x.shape[-1] // k))
+    
+        return tf.reduce_mean(y, axis=(-2,))
```
Comment (eickenberg) [inline kymatio/scattering1d/backend/tensorflow_backend.py:36]: `axis=-2`

File: kymatio/backend/tensorflow_backend.py (hunk 17 above)
Comment (janden) [reply kymatio/backend/tensorflow_backend.py:2]: Just for consistency.

File: tests/scattering1d/test_numpy_backend_1d.py [hunk 23]
```diff
@@ -0,0 +1,117 @@
+import pytest
```
Comment (eickenberg) [inline tests/scattering1d/test_numpy_backend_1d.py:1]: are these tests from this pr or from the tests refactoring pr? I don't see why the tests would change with this refactoring

File: kymatio/backend/numpy_backend.py (hunk 16 above)
Comment (janden) [reply kymatio/backend/numpy_backend.py:30]: > the self.name will be inherited?

That's probably not a good idea. (What is `backend.name` used for, anyhow?)

> will we also be having a cupy backend then?

Should be relatively straightforward if we want to go there (don't think we do, however).

File: kymatio/backend/torch_backend.py (hunk 6 above)
Comment (janden) [reply kymatio/backend/torch_backend.py:134]: Same here.

File: kymatio/backend/tensorflow_backend.py (hunk 5 above)
Comment (janden) [reply kymatio/backend/tensorflow_backend.py:45]: The second one won't work since inheritance in Python clobbers the method in the superclass (I think). How about this?
```
def contatenate(self, arrays):
    return tf.stack(arrays, dim=1)
```
The way we've set things up, the second dimension (`dim = 1`) is always the coefficient dimension, so we always want to concatenate along there, independent of the dimension of the signals. So there's only one `concatenate` that is used by all the backends.

File: tests/scattering1d/test_numpy_backend_1d.py (hunk 23 above)
Comment (MuawizChaudhary) [reply tests/scattering1d/test_numpy_backend_1d.py:1]: This is from the test refactoring PR. We should merge those as soon as possible.

File: kymatio/scattering1d/backend/numpy_backend.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (janden) [reply kymatio/scattering1d/backend/numpy_backend.py]: Yes that's what I meant to say. Pass it to the constructor in the superclass, then we don't have to assign it here.

File: kymatio/backend/torch_backend.py (hunk 6 above)
Comment (janden) [reply kymatio/backend/torch_backend.py:133]: Can't it get them from the backend? I agree that these functions belong more to the frontend, but they're currently organized into the backend, so let's keep them there.

File: tests/scattering1d/test_torch_scattering1d.py (hunk 10 above)
Comment (janden) [reply tests/scattering1d/test_torch_scattering1d.py]: Ok.

File: kymatio/scattering1d/backend/tensorflow_backend.py (hunk 21 above)
Comment (janden) [reply kymatio/scattering1d/backend/tensorflow_backend.py:6]: He means the class structure. I think it's good to have for consistency between the backends, but it's not strictly necessary. (Although it would simplify things if we have `TensorFlowBackend` inherit from `NumPyBackend`, since they use the same input check functions.)

**Comment by eickenberg** [comment]
I'd like to discuss today the differences between making changes by subclassing and making changes by passing a constructor argument.

File: kymatio/backend/tensorflow_backend.py (hunk 5 above)
Comment (janden) [reply kymatio/backend/tensorflow_backend.py:45]: 3D is a world unto itself, I'm afraid. It may warrant a special `concatenate_3d` that does something different.

File: tests/scattering1d/test_torch_backend_1d.py (hunk 9 above)
Comment (MuawizChaudhary) [reply tests/scattering1d/test_torch_backend_1d.py:179]: could you explain this for me during the meeting?

**Comment by janden** [comment]
@MuawizChaudhary Can you rebase this on top of dev? There's a lot of conflicts that need to be resolved and I don't want us to have to review it twice.

**Comment by MuawizChaudhary** [comment]
If you squashed all the changes, I'll re base it shortly!

**Comment by janden** [comment]
> If you squashed all the changes, I'll re base it shortly!

The changes in dev from the 1D test PR? Yeah they're squashed.

**Comment by janden** [comment]
Tests fail. Please fix and we'll review.

File: kymatio/scattering1d/backend/tensorflow_backend.py (hunk 21 above)
Comment (eickenberg) [reply kymatio/scattering1d/backend/tensorflow_backend.py:5]: Ah is this why I see tf backend inherit from numpy backend?
How sure are we that this works and doesn't expose a lot of numpy functionality in the tf backend?
I don't even see the input check functions work properly because I don't see where the np object is set.
Is this passing tests? Are there even tests checking these things?

File: kymatio/backend/numpy_backend.py [hunk 24]
```diff
@@ -1,73 +1,81 @@
...
-        raise RuntimeError('The inputs are not compatible for '
-                           'multiplication.')
-
-    if not _is_complex(B) and not _is_real(B):
-        raise TypeError('The second input must be complex or real.')
-
-    return A * B
+   
+class NumpyBackend:
+    def __init__(self, name='numpy'):
```
[the remark this answers is not in the document]
Comment (MuawizChaudhary) [reply kymatio/backend/numpy_backend.py:4]: We need it so that we can specify whats the name of the backend.  We dont have to if its nD numpy backend since we already specified the name here.

File: kymatio/backend/numpy_backend.py [hunk 25]
```diff
@@ -1,73 +1,81 @@
...
+        import numpy
```
[the remark this answers is not in the document]
Comment (MuawizChaudhary) [reply kymatio/backend/numpy_backend.py:5]: I believe it shouldn't trigger a numpy import unless we call the init functions. 

I thought the best idea was to import it inline, I believe that might have been what @janden wanted.

File: kymatio/backend/tensorflow_backend.py [hunk 26]
```diff
@@ -1,72 +1,43 @@
...
+    def __init__(self):
+        super(TensorFlowBackend, self).__init__(name='tensorflow')
 
-def real_check(x):
-    if not _is_real(x):
-        raise TypeError('The input should be real.')
+    def concatenate(self, arrays):
+        return tf.stack(arrays, axis=1)
 
-def _is_complex(x):
```
[the remark this answers is not in the document]
Comment (MuawizChaudhary) [reply kymatio/backend/tensorflow_backend.py:13]: This functionality is in the numpy backend.

File: kymatio/backend/tensorflow_backend.py [hunk 27]
```diff
@@ -1,72 +1,43 @@
+from .numpy_backend import NumpyBackend
 import tensorflow as tf
-import numpy as np
 
 
-def complex_check(x):
-    if not _is_complex(x):
-        raise TypeError('The input should be complex.')
+class TensorFlowBackend(NumpyBackend):
```
[the remark this answers is not in the document]
Comment (MuawizChaudhary) [reply kymatio/backend/tensorflow_backend.py:5]: We inherit for the complex and real check functions. tf and np share the same code. 

I over ride the fft functions in the 1d backend.

File: kymatio/backend/tensorflow_backend.py [hunk 28]
```diff
@@ -1,72 +1,43 @@
...
+    def __init__(self):
+        super(TensorFlowBackend, self).__init__(name='tensorflow')
```
[the remark this answers is not in the document]
Comment (MuawizChaudhary) [reply kymatio/backend/tensorflow_backend.py:7]: the line above?

File: kymatio/backend/tensorflow_backend.py [hunk 29]
```diff
@@ -1,72 +1,43 @@
...
+        if not self._is_complex(A):
+            raise TypeError('The first input must be complex.')
+    
+        if A.shape[-len(B.shape):] != B.shape[:]:
+            raise RuntimeError('The inputs are not compatible for multiplication.')
+    
+        if not self._is_complex(B) and not self._is_real(B):
+            raise TypeError('The second input must be complex or real.')
+    
+        return A * B
```
[the remark this answers is not in the document]
Comment (MuawizChaudhary) [reply kymatio/backend/tensorflow_backend.py:43]: You have a point.

File: kymatio/backend/torch_backend.py [hunk 30]
```diff
@@ -115,111 +95,116 @@ def backward(ctx, grad_output):
...
+    def __init__(self):
+        self.name = 'torch'
+ 
+    def input_checks(self, x):
+        if x is None:
+            raise TypeError('The input should be not empty.')
+    
+        self.contiguous_check(x)
+    
+    def complex_check(self, x):
```
[the remark this answers is not in the document]
Comment (MuawizChaudhary) [reply kymatio/backend/torch_backend.py:109]: I feel like this is a smart idea.

File: kymatio/backend/torch_skcuda_backend.py [hunk 31]
```diff
@@ -2,82 +2,83 @@
...
-        m, n = B.nelement() // 2, A.nelement() // B.nelement()
-        lda = m
-        ldc = m
-        incx = 1
-        handle = torch.cuda.current_blas_handle()
-        stream = torch.cuda.current_stream()._as_parameter_
-        cublas.cublasSetStream(handle, stream)
-        cublas.cublasCdgmm(handle, 'l', m, n, A.data_ptr(), lda, B.data_ptr(), incx, C.data_ptr(), ldc)
-        return C
+class TorchSKcudaBackend:
```
[the remark this answers is not in the document]
Comment (MuawizChaudhary) [reply kymatio/backend/torch_skcuda_backend.py:5]: we inherit the torch backend from the 1d torch backend when we create the TorchSKcuda 1D backend class. The torch skcuda base backend is also inherited.

[the remark this answers is not in the document]
Comment (MuawizChaudhary) [reply kymatio/backend/torch_skcuda_backend.py:5]: I recommend checking `kymatio/scattering1d/backend/torch_skcuda_backend.py`.

File: kymatio/scattering1d/backend/numpy_backend.py [hunk 32]
```diff
@@ -1,116 +1,98 @@
...
-    real_check(x)
-    return scipy.fftpack.fft(x)
-
-
-def irfft(x):
-    complex_check(x)
-    return scipy.fftpack.ifft(x).real
+        y = x.reshape(-1, k, x.shape[-1] // k)
+    
+        res = y.mean(axis=-2)
```
[the remark this answers is not in the document]
Comment (janden) [reply kymatio/scattering1d/backend/numpy_backend.py:32]: This is what was done before so it's likely correct.

File: tests/general/test_torch_backend.py [hunk 33]
```diff
@@ -1,16 +1,16 @@
...
-from kymatio.backend.torch_backend import ModulusStable, modulus
+from kymatio.backend.torch_backend import ModulusStable, TorchBackend
 
 
 def test_modulus(random_state=42):
     """
     Tests the stability and differentiability of modulus
     """
-
+    backend = TorchBackend()
```
Comment (janden) [inline tests/general/test_torch_backend.py:10]: Why not just import `backend` from `kymatio.backend.torch_backend`? It's already been instantiated and that's what the main code will use.

File: tests/scattering1d/test_tensorflow_backend_1d.py [hunk 34]
```diff
@@ -1,6 +1,6 @@
 import pytest
 import numpy as np
-
+import tensorflow as tf
```
Comment (janden) [inline tests/scattering1d/test_tensorflow_backend_1d.py:3]: Why do we need this?

File: tests/scattering1d/test_tensorflow_backend_1d.py [hunk 35]
```diff
@@ -106,14 +106,11 @@ def coefficent(n):
     y_r = (x_r * coefficents).sum(-1)
 
     z = backend.rfft(x_r)
-    # increase tolerance here as tensorflow fft is slightly inaccurate due to
```
Comment (janden) [inline tests/scattering1d/test_tensorflow_backend_1d.py:109]: Why is this going away?

File: tests/scattering1d/test_torch_backend_1d.py (hunk 9 above)
Comment (janden) [reply tests/scattering1d/test_torch_backend_1d.py:179]: We'd have a list of compatible backends and devices and loop over those. So if skcuda is available, we'd have the list: `[(torch_backend, 'cpu'), (torch_backend, 'gpu'), (torch_skcuda_backend, 'gpu')` and parametrize the tests with that. This is implemented in the 2D Torch tests. No need to do it right now, but we should probably do this in the 1D and 3D tests as well.

**Review by janden — DISMISSED** [review]
Looks good. Just a few things to address, then we can extend this to 2D and 3D.

**Comment by eickenberg** [comment]
OK, I just rebased this.

I will need some help to figure out what the intention was in some of these changes.

The global picture is that we are introducing classes for backends instead of the named tuples.

Please take a look at the changes I needed to make in my first commit on top of this to make the tests for numpy backend 2d pass: https://github.com/kymatio/kymatio/pull/631/commits/8ac8a8b42794881bb8c0aa8661bdec29aec4c482

basically I had to write a bunch of attributes to the instantiated class. I left the original commented for comparison.

Question: Why is it that we are defining half of the backend in the class and half in the 2d backend file? Is it because those parts are the specific stuff? Should we potentially be doing this using inheritance?

Happy for any input and opinions @janden @MuawizChaudhary

**Comment by eickenberg** [comment]
OK, I have kludged my way to making the tests pass.

For the specific backends I would suggest using inheritance rather than writing properties, since that is not much better than the named tuples we had before.

Please let me know if you think that makes sense and I'll make those changes. Ideally we get this one merged before the sprint (or at the latest at the beginning of it).

**Comment by janden** [comment]
Still fails, but for legit reasons (torch_skcuda backend has not been updated to the new architecture).

**Comment by janden** [comment]
So tests pass now on Jenkins. Since it's been a while, I want to go over everything again to see what needs to be done before merging.

**Comment by eickenberg** [comment]
Awesome. I will read through the updated code now.
One thing that I have had simmering for a bit is the instantiation. Do we really need to instantiate these backends?
If not, we could set all the methods to `staticmethod` and remove the `self`objects because they are not needed

**Comment by edouardoyallon** [comment]
the fft failure for python 3.6 *only*, with conda and tf sounds like some future promising fun!

File: kymatio/backend/numpy_backend.py (hunk 25 above)
[the remark this answers is not in the document]
Comment (eickenberg) [reply kymatio/backend/numpy_backend.py:5]: Yeah I think nothing is really stopping us from using a global import.

I'd generally also like to get away as much as possible from making an instance. Ideally all functions are static on the class and we use inheritance to pass down those static functions

File: kymatio/backend/tensorflow_backend.py (hunk 29 above)
[the remark this answers is not in the document]
Comment (eickenberg) [reply kymatio/backend/tensorflow_backend.py:44]: need to check whether we can unify cdgmm across dimensionalities

File: kymatio/backend/torch_backend.py (hunk 30 above)
[the remark this answers is not in the document]
Comment (eickenberg) [reply kymatio/backend/torch_backend.py:109]: I guess this is the base class actually.
But we might be able to use one more level of inheritance to deal with the two different complex number formulations

File: kymatio/backend/torch_backend.py [hunk 36]
```diff
@@ -115,111 +95,117 @@ def backward(ctx, grad_output):
...
 
-        Parameters
-        ---------
-        x : tensor
-            Complex torch tensor.
+    def complex_contiguous_check(self, x):
+        self.complex_check(x)
+        self.contiguous_check(x)
+    
+    def contiguous_check(self, x):
```
Comment (eickenberg) [inline kymatio/backend/torch_backend.py:121]: yeah I'm not happy with these unused `self` objects

File: kymatio/frontend/numpy_frontend.py [hunk 37]
```diff
@@ -11,7 +11,7 @@ def scattering(self, x):
     def __call__(self, x):
         """This method is an alias for `scattering`."""
 
-        input_checks(x)
+        self.backend.input_checks(x)
```
[the remark this answers is not in the document]
Comment (eickenberg) [reply kymatio/frontend/numpy_frontend.py:14]: you mean because `scattering` assumes checks to have been done?

File: tests/scattering1d/test_tensorflow_backend_1d.py (hunk 35 above)
Comment (eickenberg) [reply tests/scattering1d/test_tensorflow_backend_1d.py:109]: yeah why?

File: tests/scattering1d/test_torch_backend_1d.py (hunk 9 above)
Comment (eickenberg) [reply tests/scattering1d/test_torch_backend_1d.py:179]: good idea

**Comment by eickenberg** [comment]
@janden is it OK if I go through the backends and remove any unnecessary use of `self`in function definitions? I.e. functions that do not depend on state should not need a `self` object as input

File: kymatio/backend/numpy_backend.py (hunk 25 above)
[the remark this answers is not in the document]
Comment (janden) [reply kymatio/backend/numpy_backend.py:5]: > Yeah I think nothing is really stopping us from using a global import.

Agree. That's how it is for the other backends, as far as I can tell.

> I'd generally also like to get away as much as possible from making an instance. Ideally all functions are static on the class and we use inheritance to pass down those static functions

I remember inheriting static functions as being slightly tricky to deal with when it comes to inheritance, but this may be a false memory. Overall, I think it's simple enough to have these singleton instances since it allows us to store some internal state (like these module references).

[the remark this answers is not in the document]
Comment (eickenberg) [reply kymatio/backend/numpy_backend.py:5]: why can't we just write those into the class?
I'm happy to keep any state that is necessary, but would avoid it if it's easy

**Comment by janden** [comment]
@eickenberg So all the backends have been refactored now to have either class methods or stack methods. There are no object instantiations (as far as I can tell). Please have a look and let me know what you think.

File: kymatio/scattering3d/backend/tensorflow_backend.py [hunk 38]
```diff
@@ -7,116 +7,109 @@
...
-backend.name = 'tensorflow'
-backend.cdgmm3d = cdgmm3d
-backend.rfft = rfft
-backend.ifft = ifft
-backend.modulus = Modulus()
-backend.modulus_rotation = modulus_rotation
-backend.compute_integrals = compute_integrals
-backend.concatenate = concatenate
+
+backend = TensorFlowBackend3D()
```
Comment (eickenberg) [inline kymatio/scattering3d/backend/tensorflow_backend.py:115]: instantiation not necessary, correct?

File: tests/general/test_torch_backend.py (hunk 33 above)
Comment (eickenberg) [reply tests/general/test_torch_backend.py:10]: especially now, since it's actually not instantiated

Comment (eickenberg) [reply tests/general/test_torch_backend.py:10]: though that said, it might be more modular to do it like above, just without the parentheses

File: tests/scattering1d/test_numpy_backend_1d.py (hunk 23 above)
Comment (eickenberg) [reply tests/scattering1d/test_numpy_backend_1d.py:1]: hmm,  i believe this is merged, so I'm surprised to see the diff below

**Review by eickenberg — COMMENTED** [review]
This looks close to mergeable. At this point I think if we address all the remaining comments, we should be good to go

**Review by eickenberg — APPROVED** [review]
OK, I think I'm done for today.

I have addressed a number of outstanding code comments in the PR.

The remaining ones I am not sure how to address, but I have the feeling they are minor.

I did an interactive rebase in order to consolidate @MuawizChaudhary 's summer work with my commits from last week so that in aggregate the tests pass on that commit. Just in case this wasn't a good idea, I have the unsquashed branch fully backed up on my computer. You can also reconstruct the previous status quo by using what you have on your machine and adding my most recent commits on top.

There is one remaining puzzle: I found (and had commented on in the PR at some point) several changes that looked like they were from a previous test refactoring project. I had thought that these might be duplicates, but for most of them, I can't find a trace of a second version of them anywhere in the history. If I am right, then we should make sure to keep these commits, because they do work on the tests.

From my side this can be merged as is, since merging will just add the test commits, and the remaining comments are minor. However, it is also possible to address the comments and split off the TST commits in order to make this PR only about the backend refactor.

I am fine with either. Since it's the last sprint day tomorrow, maybe we can get this in and then try to add the jax backends?

**Comment by eickenberg** [comment]
OK, as discussed with @janden, I squashed all the straggling TST commits together and hilariously all those commits boiled down to these very few lines removing the tolerance from tensorflow fft test https://github.com/kymatio/kymatio/pull/631/commits/a6dee636542e6f8f8ffa739b8be63f49f39fb857

I will now proceed to remove this commit, because we don't actually want that change.

**Comment by janden** [comment]
:+1:
